enhancements/network/on-prem-dual-stack-vips.md

```diff
@@ -0,0 +1,421 @@
+the second VIP if it is not needed.
+
+## Alternatives
+
+Implement this somewhat like the
+[existing workaround](https://github.com/yboaron/kepalived-v6) - create a
+separate instance of keepalived that is responsible for ipv6 only. This
+results in some duplication of containers and configs, but it would probably
+simplify the code changes.
+
+### Alternative install-config layouts
+
+#### Just add v6 versions of the VIP fields
+```yaml
+platform:
+  baremetal:
+    apiVIP: 192.168.111.5
+    apiVIPv6: f00::5
+    ingressVIP: 192.168.111.4
+    ingressVIPv6: f00::4
+```
+This is a minimal change, but it leaves us with somewhat of a problem in that
+we can't assume the old fields always contain a v4 address since previous
+releases allowed either v4 or v6. We may be able to deal with that by
+detecting the appropriate ip version for the old fields and populating new
+fields on the api side. E.g. if apiVIP is v4, we populate
+`platformStatus.apiServerInternalIPv4` and if it is v6 we populate
+`apiServerInternalIPv6` instead. That way the only place we need to worry about
+the difference is in the installer. We would need a validation to ensure a v6
+VIP isn't specified in both fields.
+
+Note that this assumes we replace the existing api fields with two new ones
+per VIP and do not reuse the existing ones at all. I think we're going to want
+to do that regardless of what install-config layout we use so we don't need
+logic on the backend in machine-config-operator and baremetal-runtimecfg to
+handle older cluster configs.
+
+#### Deprecate the old fields and add new v4 and v6 fields
+```yaml
+platform:
+  baremetal:
+    # Deprecated
+    apiVIP: 192.168.111.5
+    # Deprecated
+    ingressVIP: 192.168.111.4
+
+    apiVIPv4: 192.168.111.5
+    apiVIPv6: f00::5
+    ingressVIPv4: 192.168.111.4
+    ingressVIPv6: f00::4
+```
+This is more of a 1-1 mapping to what I anticipate the platformStatus api
+will look like. We would still need compatibility logic to handle old-style
+install-configs that set the deprecated fields, but having no overlap between
+the old and new-style configs might simplify the logic somewhat.
+
+#### Add plural fields that take a list of VIPs
+```yaml
+platform:
+  baremetal:
+    # Deprecated
+    apiVIP: 192.168.111.5
+    # Deprecated
+    ingressVIP: 192.168.111.4
+
+    apiVIPs:
+    - 192.168.111.5
+    - f00::5
+    ingressVIPs:
+    - 192.168.111.4
+    - f00::4
+```
+This is essentially a variation on the previous option where the two separate
+fields per VIP are combined into a list field with a length of up to 2. In
+this case we'd need some logic to determine whether the VIPs in the list are
+v4 or v6, but I expect that would only be a little more complicated than the
+other options. This has the advantage of being extensible if we ever wanted
+to add more addresses, but I admit to having a hard time coming up with a use
+case where more than an ipv4 and ipv6 address would be needed.
```

Comment:
I could imagine an on-prem scenario where there is both an application network plus an independent management network. Since we blur the lines of the k8s API, using it both for infrastructure management and application management, it would be necessary to have the API service listening on both networks.

Likewise with RHACM you could imagine a hub-and-spoke model where the hub (management) cluster has special access to multiple otherwise-separate networks so it can manage infrastructure in multiple locations / facilities / regions etc. It would benefit from having VIPs on those multiple networks.

I don't know that there's been any direct demand for the above patterns, but they seem plausible.